### PR TITLE
[Inference]Add CUDA KVCache Kernel

### PR DESCRIPTION
**Pull request by yuehuayingxueluo**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.
pytest:
image
ops benchmark:
image
image
model benchmark:
| bsz | in_len | out_len | Throughput (tokens/sec) | 
|-----|--------|---------|--------------------------|
| 16  | 128    | 128     | 1483.78-> 1647.18 | 
| 32  | 128    | 128     | 2730.67 -> 3000.20 | 
| 64  | 128    | 128     | 4851.50 -> 4954.12 | 
| 16  | 512    | 128     | 1319.74 -> 1379.97 |     
| 32  | 512    | 128     | 2020.19 -> 2053.62 | 
| 64  | 512    | 128     | 2676.58 -> 2689.92 |     
| 16  | 1024   | 128     | 1029.48 -> 1024.90 |      
| 32  | 1024   | 128     | 1345.16 -> 1347.42 |     
| 64  | 1024   | 128     | 1603.75 -> 1610.99 |   

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/inference/modeling/models/nopadding_llama.py [hunk 1]
```diff
@@ -22,6 +23,14 @@
 )
 from colossalai.logging import get_dist_logger
 
+try:
+    from colossalai._C import inference_ops_cuda as inference_ops
+except:
+    inference_ops = None
+
+if inference_ops is None:
+    inference_ops = InferenceOpsLoader().load()
```
Comment (FrankLeeeee) [inline colossalai/inference/modeling/models/nopadding_llama.py:32]: `InferenceOpsLoader.load()` should be enough, why do you still need to check `colossalai._C`?

File: colossalai/inference/modeling/models/nopadding_llama.py [hunk 2]
```diff
@@ -74,6 +83,9 @@ def llama_model_forward(
     sequence_lengths = batch.get_sequence_lengths()
     batch_size = batch.current_batch_size
     kv_seq_len = sequence_lengths.max().item()
+    use_cuda = True
+    if batch_size >= 32 and kv_seq_len > 512:
+        use_cuda = False
```
Comment (FrankLeeeee) [inline colossalai/inference/modeling/models/nopadding_llama.py:88]: Write some comments to explain the reason behind this if statement.

File: colossalai/inference/modeling/models/nopadding_llama.py [hunk 3]
```diff
@@ -107,6 +119,7 @@ def llama_model_forward(
             output_tensor=output_tensor,
             norm_output=norm_output,
             sm_scale=sm_scale,
+            use_cuda=use_cuda,
```
Comment (FrankLeeeee) [inline colossalai/inference/modeling/models/nopadding_llama.py:122]: `use_cuda` is rather confusing to me. Shouldn't it be always running on GPU? Do you mean `use_fused_cuda_kernel`?

File: examples/inference/benchmark_ops/benchmark_kv_cache_memcopy.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+except ImportError:
+    print("please install triton from https://github.com/openai/triton")
+
+try:
+    from colossalai._C import inference_ops_cuda as inference_ops
+except:
+    inference_ops = None
+
+if inference_ops is None:
+    inference_ops = InferenceOpsLoader().load()
```
Comment (FrankLeeeee) [inline examples/inference/benchmark_ops/benchmark_kv_cache_memcopy.py:20]: Same here, just use `InferenceOpsLoader().load()`

File: extensions/cuda_extension.py [hunk 5]
```diff
@@ -1,7 +1,10 @@
 import os
+import time
 from abc import abstractmethod
+from pathlib import Path
 from typing import List
 
+from .base_extension import _Extension
```
[the remark this answers is not in the document]
Comment (yuehuayingxueluo) [reply extensions/cuda_extension.py:7]: When I execute the code, I encounter errors indicating that 'time', 'Path', and '_Extension' are undefined. Therefore, I have added the import statements here.

File: colossalai/inference/modeling/models/nopadding_llama.py (hunk 1 above)
Comment (yuehuayingxueluo) [reply colossalai/inference/modeling/models/nopadding_llama.py]: Done

File: colossalai/inference/modeling/models/nopadding_llama.py (hunk 2 above)
Comment (yuehuayingxueluo) [reply colossalai/inference/modeling/models/nopadding_llama.py]: Done

File: colossalai/inference/modeling/models/nopadding_llama.py (hunk 3 above)
Comment (yuehuayingxueluo) [reply colossalai/inference/modeling/models/nopadding_llama.py]: Done

File: extensions/cuda_extension.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (FrankLeeeee) [reply extensions/cuda_extension.py:7]: ok

**Review by FrankLeeeee — APPROVED** [review]

**Pull request merged**